pkg/controller/worker/genericactuator/actuator_restore.go

```diff
@@ -0,0 +1,162 @@
+		return errors.Wrap(err, "failed to restore the machine deployment config")
+	}
+
+	return nil
+}
+
+func (a *genericActuator) addStateToMachineDeployment(ctx context.Context, worker *extensionsv1alpha1.Worker, wantedMachineDeployments workercontroller.MachineDeployments) error {
+	// Make a deep copy to ensure that no one will change the state during the restoration
+	workerCopy := worker.DeepCopy()
+
+	if workerCopy.Status.State == nil || len(workerCopy.Status.State.Raw) <= 0 {
+		return nil
+	}
+
+	// Parse the worker state to MachineDeploymentStates
+	workerState := make(map[string]*workercontroller.MachineDeploymentState)
+	if err := json.Unmarshal(workerCopy.Status.State.Raw, &workerState); err != nil {
+		return err
+	}
+
+	// Attach the parsed MachineDeploymentStates to the wanted MachineDeployments
+	for index, wantedMachineDeployment := range wantedMachineDeployments {
+		wantedMachineDeployments[index].State = workerState[wantedMachineDeployment.Name]
+	}
+
+	return nil
+}
+
+func (a *genericActuator) deployMachineSetsAndMachines(ctx context.Context, worker *extensionsv1alpha1.Worker, wantedMachineDeployments workercontroller.MachineDeployments) error {
+	for _, wantedMachineDeployment := range wantedMachineDeployments {
+		if wantedMachineDeployment.State == nil || wantedMachineDeployment.State.MachineSet == nil {
+			continue
+		}
+
+		machineSet := wantedMachineDeployment.State.MachineSet
+
+		// Create the MachineSet if not already exists. We do not care about the MachineSet status
+		// because the MCM will update it
+		if err := a.client.Create(ctx, machineSet); err != nil && !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+
+		// Deploy each machine owned by the MachineSet which was restored above
+		for _, machine := range wantedMachineDeployment.State.Machines {
+			// Create the machine if it not exists already
+			err := a.client.Create(ctx, &machine)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return err
+			}
+
+			// Attach the Shoot node to the Machine status
+			node := machine.Status.Node
+			if err := a.updateStatus(ctx, &machine, apierrors.IsAlreadyExists(err), func() error {
+				machine.Status.Node = node
+				return nil
+			}); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// alreadyExists must be returned from from apierrors.IsAlreadyExists(err) where err commes from client.Create.
+// This function is needed because when you call client.Create and it successed if you immediately call TryUpdateStatus
+// it will return an error stating that the Object does not exists.
+// This bug comes form the fact that exponentialBackoff in TryUpdateStatus does not retry if client.Get returns error
+// so the function exits immediately with error
+func (a *genericActuator) updateStatus(ctx context.Context, obj runtime.Object, alreadyExists bool, transform func() error) error {
+	if alreadyExists {
```

Comment:
No we can't. In `TryUpdateStatus`  first we call client.Get(...) which due to some cashes return ErrorNotFound.  The `exponentialBackoff` function does not wait if error occurs and returns it immediately. Thus  if call `client.Create(...)` and then immediately call `TryUpdateStatus` you get an error. I have wasted one day debugging this.

Reply:
Hm, ok, then I would still try to retry instead of explicitly handling the `alreadyExists` error that may occur due to not-up-to-date caches.

Why not simply retrying the `a.client.Status().Update()` call for a few seconds using the [`retry.Until()`](https://github.com/gardener/gardener-extensions/blob/c9d757465d14ed7bbfcb7e823bce0e3786c63810/pkg/terraformer/waiter.go#L31-L46) function?

Reply:
I will use retry.Until() with TryUpdateStatus